docs/development/design/future-builder.rst

```diff
@@ -0,0 +1,230 @@
+These are some examples where this is useful:
+
+- User wants to pass an extra argument to ``sphinx-build``
+- Project requires to execute a command *before* building
+- User has a personal/private PyPI URL
+- etc
+
+.. code:: yaml
+
+   # .readthedocs.yaml
+   build:
+     builder: 2
+     jobs:
+       pre_checkout:
+       checkout: git clone --branch main https://github.com/readthedocs/readthedocs.org
+       post_checkout:
+       pre_create_environment:
+       create_environment: python -m virtualenv venv
+       post_create_environment:
+       pre_install:
+       install: pip install -r requirements.txt
+       post_install:
+       pre_build:
+       build:
+         html: sphinx-build -T -j auto -E -b html -d _build/doctrees -D language=en . _build/html
+         pdf: latexmk -r latexmkrc -pdf -f -dvi- -ps- -jobname=test-builds -interaction=nonstopmode
+         epub: sphinx -T -j auto -b epub -d _build/doctrees -D language=en . _build/epub
+       post_build:
+       pre_metadata:
+       metadata: ./metadata_sphinx.py
+       post_medatada:
+
+
+.. note::
+
+   *All these commands* are executed passing all the exposed environment variables.
+
+If the user only provides a subset of these jobs, we ran our default commands if the user does not provide them
+(see :ref:`Step ran by the builder`).
+For example, the following YAML is enough when the project requires running Doxygen as a pre-build step:
+
+.. code:: yaml
+
+   # .readthedocs.yaml
+   build:
+     builder: 2
+     jobs:
+       # https://breathe.readthedocs.io/en/latest/readthedocs.html#generating-doxygen-xml-files
+       pre_build: cd ../doxygen; doxygen
+
+
+``build.commands``
+~~~~~~~~~~~~~~~~~~
+
+It allows users to have full control over the commands executed in the build process.
+These are some examples where this is useful:
+
+- project with a custom build process that does map ours
+- specific requirements that we can't/want to cover as a general rule
+- build documentation with a different tool than Sphinx
+
+
+.. code:: yaml
+
+   # .readthedocs.yaml
+   build:
+     builder: 2
+     commands:
+       - git clone --branch main https://github.com/readthedocs/readthedocs.org
+       - pip install -r requirements.txt
+       - sphinx-build -T -j auto -E -b html -d _build/doctrees -D language=en . _build/html
+       - ./metadata.py
+
+
+Intermediate steps for rollout
+------------------------------
+
+#. Remove all the exposed data in the ``conf.py.tmpl`` file and move it to ``metadata.yaml``
+#. Define structure required for ``metadata.yaml`` as contract
+#. Define the environment variables required (e.g. some from ``html_context``) and execute all commands with them
+#. Build documentation using this contract
+#. Leave ``readthedocs-sphinx-ext`` as the only package installed and extension install in ``conf.py.tmpl``
+#. Add ``build.builder: 2`` config without any *magic*
```

Comment:
I'm not really sold on the idea of versioning this, as it would be confusing having two versions (one for the config file and other for how the docs are build?). Also, I'd call it `build.version` as `builder: 2` looks like you are requiring more builders.

Reply:
Yeah, I understand that it may be confusing having the global `version` and `build.version`. However, we need a way to differentiate the "builder that does some magic" (current behavior) from the "builder that does nothing". What do you think could be better for this?